parse/planner/logical/rewriter.go

```diff
@@ -0,0 +1,484 @@
+	// and any expression acting on a plan will be called before visiting the plan.
+	// If false, the callback will be called before visiting children,
+	// and any expression acting on a plan will be called after visiting the plan.
+	// ! IMPORTANT: If this is set to true, then the callback functions must always return false.
+	// ! I am not sure if we need this anymore.
+	CallbackAfterVisit bool
+	// PostOrderVisit determines the order in which fields are visited.
+	// If visitng in post order, then children are visited first, then the parent.
+	// For example, for a Project node, if PostOrderVisit is true, then the child
+	// is visited first, then the expressions.
+	// If PostOrderVisit is false, then the expressions are visited first, then the child.
+	PostOrderVisit bool
+}
+
+// Rewrite rewrites a logical plan using the given configuration.
+// It returns the rewritten plan, but it also modifies the original plan in place.
+// The returned plan should always be used.
+func Rewrite(node LogicalNode, cfg *RewriteConfig) (lp LogicalNode, err error) {
+	// defer func() {
+	// 	if r := recover(); r != nil {
+	// 		err2, ok := r.(error)
+	// 		if !ok {
+	// 			err = fmt.Errorf("%v", r)
+	// 		} else {
+	// 			err = err2
+	// 		}
+	// 	}
+	// }()
+
+	v := &rewriteVisitor{
+		exprCallback:       cfg.ExprCallback,
+		planCallback:       cfg.PlanCallback,
+		scanSourceCallback: cfg.ScanSourceCallback,
+		callbackFuncLast:   cfg.CallbackAfterVisit,
+		postOrderVisit:     cfg.PostOrderVisit,
+	}
+	if v.exprCallback == nil {
+		v.exprCallback = func(e Expression) (Expression, bool, error) {
+			return e, true, nil
+		}
+	}
+	if v.planCallback == nil {
+		v.planCallback = func(p Plan) (Plan, bool, error) {
+			return p, true, nil
+		}
+	}
+	if v.scanSourceCallback == nil {
+		v.scanSourceCallback = func(s ScanSource) (ScanSource, bool, error) {
+			return s, true, nil
+		}
+	}
+
+	return node.Accept(v).(LogicalNode), nil
+}
+
+// rewriteVisitor is a visitor that can be used to rewrite a logical plan.
+// For all visits, functions can be given to determine the order in which
+// fields are visited in. They should be passed in post-order (e.g. in the same
+// order as they would need to be visited to evaluate the relation).
+type rewriteVisitor struct {
+	// exprCallback is the function that will be called on each expression
+	exprCallback func(Expression) (Expression, bool, error)
+	// planCallback is the function that will be called on each plan
+	planCallback func(Plan) (Plan, bool, error)
+	// scanSourceCallback is the function that will be called on each scan source
+	scanSourceCallback func(ScanSource) (ScanSource, bool, error)
+	// if true, the callback will be called before visiting children
+	// if false, the callback will be called after visiting children
+	callbackFuncLast bool
+	// if true, then fields are visited in post order
+	// if false, then fields are visited in pre order
+	postOrderVisit bool
```

Comment:
I find this configuration confusing. To me, `callbackFuncLast` decides whether it's preOrder or postOrder.

The 'order' that 'postOrderVisit' controls doesn't matter.

Reply:
I agree. I removed them from the rewriter for now since they are unused, but left the internals as there will still be a lot of rewriting I do.